File: Coding-Interviews/31-栈的压入、弹出序列.cpp
#include <stack>
using namespace std;
class Solution
{
public:
    bool validateStackSequences(vector<int>& pushed, vector<int>& popped)
    {
        stack<int> s;
        int        cur1 = 0;
        int        cur2 = 0;
        while (cur2 < popped.size()) {
            while (s.empty() || s.top() != popped[cur2]) {
                if (cur1 >= pushed.size())
                    return false;
                s.push(pushed[cur1]);
                cur1++;
            }
            s.pop();
            cur2++;
        }
        return true;
    }
};

File: Coding-Interviews/31-栈的压入、弹出序列.py
class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        real = []
        idxPush = 0
        for i in popped:
            while not real or real[-1] != i:
                if idxPush < len(pushed):
                    real.append(pushed[idxPush])
                    idxPush += 1
                else:
                    return False
            real.pop()
        return True
